src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java

```diff
@@ -436,6 +436,72 @@ public final Response getQuizAttempt(@Context final HttpServletRequest httpServl
         }
     }
 
+    /**
+     * Mark a QuizAttempt as complete (or incomplete).
+     *
+     * Only owner and group managers for the quiz assignment can mark it incomplete.
+     * If it is a self-taken quiz, you cannot mark it incomplete, as that would make it easier to tweak individual
+     * questions to find the right answers.
+     *
+     * @param httpServletRequest
+     *            - so that we can extract user information.
+     * @param quizAttemptId
+     *            - the ID of the quiz the user wishes to attempt.
+     * @param markIncomplete
+     *            - optional boolean for group manager to mark incomplete.
+     * @return Confirmation or an error.
+     */
+    @POST
+    @Path("/attempt/{quizAttemptId}/complete")
+    @Produces(MediaType.APPLICATION_JSON)
+    @GZIP
+    @ApiOperation(value = "Mark a QuizAttempt as complete (or incomplete).")
+    public final Response completeQuizAttempt(@Context final HttpServletRequest httpServletRequest,
+                                              @PathParam("quizAttemptId") final Long quizAttemptId,
+                                              @FormParam("markIncomplete") final Boolean markIncomplete) {
```

Comment:
I notice I mised a few already, but `@FormParam` isn't very JSON RESTful, I don't think? I've pushed the bounds before now by allowing a [top-level list as JSON](https://github.com/isaacphysics/isaac-api/blob/v2.15.2/src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java#L532-L535) or [a single string value as JSON](https://github.com/isaacphysics/isaac-api/blob/v2.15.2/src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java#L1599-L1601), since we have never had `application/x-www-form-urlencoded` in requests to the API before I don't think. You might also find it tricky to make Axios, our frontend request library to make non-JSON requests although I have never tried.

Reply:
I'm not sure if I'm 100% following.

I was assuming I would call this with ordinary POST params, not JSON.

But I think you're saying there is generally a preference for a packaged JSON object? If so, then I can modify APIs to do that instead. (Although this one I'd probably just split into two endpoints.)

Reply:
As far as I am aware, every POST endpoint we have consumes JSON, yes. I think the original way was to make a DTO for each endpoint, but those two places above are examples of being loose about what sort of JSON value is allowed at top-level, and there are [other examples with lists too](https://github.com/isaacphysics/isaac-api/blob/v2.15.2/src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java#L372-L377).
But I have asked @mlt47 for an opinion when he's out of a meeting!

Reply:
I don't think it matters too much, but yeah sticking to JSON or a query param is what we've generally done

Reply:
I've split into two endpoints, which I think actually makes them clearer anyway.